File: utils/nameResolution.js
import { getUserRecord } from '../tools/sheets.js';
import { getSlackName } from './slackUtils.js';

const nameCache = {};

function nk(s) {
  return s.toLowerCase().replace(/[^a-z0-9]/g, '');
}

export async function getPreferredName(slackId) {
  const row = await getUserRecord(slackId);
  if (!row) return null;

  // Claves típicas
  let pref = row['Name (pref)'];
  if (typeof pref === 'string' && pref.trim()) return pref.trim();

  let first = row['Name (first)'];
  if (typeof first === 'string' && first.trim()) return first.trim();

  // Normalizadas
  const rn = Object.fromEntries(Object.entries(row).map(([k, v]) => [nk(k), v]));
  pref = rn.namepref;
  first = rn.namefirst || rn.firstname;
  if (typeof pref === 'string' && pref.trim()) return pref.trim();
  if (typeof first === 'string' && first.trim()) return first.trim();

  return null;
}

export async function resolveName(slackId) {
  if (nameCache[slackId]) return nameCache[slackId];

  let name = await getSlackName(slackId);
  const pref = await getPreferredName(slackId);
  if (pref) name = pref;

  if (name) nameCache[slackId] = name;
  return name;
}

File: utils/name_resolution.py
from tools.sheets import get_user_record, get_preferred_name as sheets_get_preferred_name
from utils.slack_utils import get_slack_name

_name_cache = {}

def _nk(s: str) -> str:
    """Normaliza nombres de columnas: minúsculas y alfanumérico."""
    return "".join(ch for ch in s.lower() if ch.isalnum())

def get_preferred_name(slack_id):
    """Devuelve Name (pref) o Name (first) desde Sheets."""
    row = get_user_record(slack_id)
    if not row:
        return None
    # intenta claves típicas
    if isinstance(row.get("Name (pref)"), str) and row["Name (pref)"].strip():
        return row["Name (pref)"].strip()
    if isinstance(row.get("Name (first)"), str) and row["Name (first)"].strip():
        return row["Name (first)"].strip()
    # intenta variantes normalizadas
    rn = { _nk(k): v for k, v in row.items() }
    pref = rn.get("namepref")
    first = rn.get("namefirst") or rn.get("firstname")
    if isinstance(pref, str) and pref.strip():
        return pref.strip()
    if isinstance(first, str) and first.strip():
        return first.strip()
    return None

def resolve_name(slack_id):
    if slack_id in _name_cache:
        return _name_cache[slack_id]
    name = get_slack_name(slack_id)     # rápido
    pref = get_preferred_name(slack_id) # si hay en Sheet, sobrescribe
    if pref:
        name = pref
    if name:
        _name_cache[slack_id] = name
    return name
